# Desafio control de flujo : Pablo Veloz

# Desafío 1
Generar un par de arrays ficticios con numpy
tip: utilice la función linspace para generar un array entre 1 y 50, y un array entre 50 y 150.

In [3]:
import numpy as np
import pandas as pd
import sys

* array entre 1 y 50

In [4]:
np.linspace(1, 50, num=50)

array([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12., 13.,
       14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25., 26.,
       27., 28., 29., 30., 31., 32., 33., 34., 35., 36., 37., 38., 39.,
       40., 41., 42., 43., 44., 45., 46., 47., 48., 49., 50.])

* array entre 50 y 150.

In [5]:
np.linspace(50, 150, num=51)

array([ 50.,  52.,  54.,  56.,  58.,  60.,  62.,  64.,  66.,  68.,  70.,
        72.,  74.,  76.,  78.,  80.,  82.,  84.,  86.,  88.,  90.,  92.,
        94.,  96.,  98., 100., 102., 104., 106., 108., 110., 112., 114.,
       116., 118., 120., 122., 124., 126., 128., 130., 132., 134., 136.,
       138., 140., 142., 144., 146., 148., 150.])

# Desafío 2
* Ejecute un loop que devuelva si el número en el primer array es par o impar
tip: utilice la función módulo para encontrar si el número es divisible por 2.

In [7]:
array1=np.linspace(1, 50, num=50)
for i in array1:
    if i%2==0:
        print("par")
    else:
        print("impar")

impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par
impar
par


# Desafío 3
* Genere un loop con el segundo array que cuente las siguientes condiciones
Si el número es divisible por 2 o 3
Si el número es divisible por 2 y 3
Si el número es divisible por 3 pero no por 2
si el número no es divisible por 2 ni 3

In [19]:
cont_23_o=0
cont_23_y=0
cont_div3_no2=0
cont_no_div3_div2=0
array2=np.linspace(50, 100, num=51)
for j in array2:
    k=int(j)
    if k%2==0 or k%3==0:
        cont_23_o+=1
    if k%2==0 and k%3==0:
        cont_23_y+=1
    if k%3==0 and k%2!=0:
        cont_div3_no2+=1
    if k%3!=0 and k%2!=0:
        cont_no_div3_div2+=1
print("Cantidad de números divisibles por 2 o 3: ",cont_23_o)
print("Cantidad de números números divisibles por 2 y 3: ",cont_23_y)
print("Cantidad de números divisibles por 3 pero no por 2: ",cont_div3_no2)
print("Cantidad de números que no son divisibles por 2 ni 3: ",cont_no_div3_div2)

Cantidad de números divisibles por 2 o 3:  35
Cantidad de números números divisibles por 2 y 3:  8
Cantidad de números divisibles por 3 pero no por 2:  9
Cantidad de números que no son divisibles por 2 ni 3:  16


# Desafío 4
* Corrección de errores
for i in range(100):
print(I**2)

Error: El error corresponde a utilizar en el ciclo for el contador con minuscula y al momento de mostrarlo lo genera con Mayuscula y eso en python lo considera como variables de distinto valor y por lo tanto señala que la variable no esta definida al momento de printear

correccion: Debe cambiar a minuscula la i dentro del print

# Desafío 5
* Utilizando la misma base flights.csv hacer un loop y clasificar los meses con una cantidad de pasajeros menor a la media
* Para ello, generen un nuevo objeto que represente la media de passengers .
* Generen una columna en la base de datos que se llame underperforming y asígnele 0.
* Ejecuten un loop que recorra cada observación de passengers , donde si la observación esmenor a la media de passengers se le asigne a underperforming un 1.
* Para asignar unos en la columna underperforming ya creada, utilice la función set_value .

In [21]:
df = pd.read_csv("flights.csv")

* Para ello, generen un nuevo objeto que represente la media de passengers .

In [23]:
df["passengers"].mean()

280.2986111111111

* Utilizando la misma base flights.csv hacer un loop y clasificar los meses con una cantidad de pasajeros menor a la media

In [22]:
for rowname,rowseries in df.iterrows():
    if (rowseries["passengers"] <= df["passengers"].mean()):
        print(rowseries["month"])

January
February
March
April
May
June
July
August
September
October
November
December
January
February
March
April
May
June
July
August
September
October
November
December
January
February
March
April
May
June
July
August
September
October
November
December
January
February
March
April
May
June
July
August
September
October
November
December
January
February
March
April
May
June
July
August
September
October
November
December
January
February
March
April
May
June
September
October
November
December
January
February
March
April
May
October
November
December
February
November


* Generen una columna en la base de datos que se llame underperforming y asígnele 0.

In [24]:
df["underperforming"]="0"

* Ejecuten un loop que recorra cada observación de passengers , donde si la observación es menor a la media de passengers se le asigne a underperforming un 1.
* Para asignar unos en la columna underperforming ya creada, utilice la función set_value .

In [66]:
j=0
for rowname,rowseries in df.iterrows():
    if (rowseries["passengers"] <= df["passengers"].mean()):
        df.set_value(j, 'underperforming', 1)
        
    else:
        df.set_value(j, 'underperforming', 0)
    j+=1

print(df)

       year      month  passengers  outlier  underperforming
0    1949.0    January       112.0      1.0              1.0
1    1949.0   February       118.0      1.0              1.0
2    1949.0      March       132.0      1.0              1.0
3    1949.0      April       129.0      1.0              1.0
4    1949.0        May       121.0      1.0              1.0
5    1949.0       June       135.0      1.0              1.0
6    1949.0       July       148.0      1.0              1.0
7    1949.0     August       148.0      1.0              1.0
8    1949.0  September       136.0      1.0              1.0
9    1949.0    October       119.0      1.0              1.0
10   1949.0   November       104.0      1.0              1.0
11   1949.0   December       118.0      1.0              1.0
12   1950.0    January       115.0      1.0              1.0
13   1950.0   February       126.0      1.0              1.0
14   1950.0      March       141.0      1.0              1.0
15   1950.0      April  

C:\Users\proveloz\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  after removing the cwd from sys.path.
C:\Users\proveloz\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  import sys


# Desafío 6
* Hacer un loop que clasifique los meses donde la cantidad de pasajeros se escapa de la tendencia.
* Para ello, genere dos objetos que guarden la media general y la desviación estandar general de passengers .
* Genere una nueva columna en la tabla de datos que se llame outlier y asignele 0.
* Ejecuten un loop que recorra cada observación de passengers , donde si la observación se escapa de la tendencia sea 1, de lo contrario 0.
* Para clasificar los casos que se escapen la tendencia, la observación debe satisfacer una de las siguientes condiciones:
* La observación debe ser menor a la media menos la desviación estándar, o
* La observación debe ser mayor a la media más la desviación estándar.
* ¿Cuántas observaciones se pueden clasificar como casos extremos?

In [39]:
df_media_pasajeros = df["passengers"].mean()
print("El promedio de pasajeros es: ",df_media_pasajeros)
df_std_pasajeros = df["passengers"].std()
print("La desviación estandar es: ",df_std_pasajeros)
delta_menos=df_media_pasajeros - df_std_pasajeros
delta_mas=df_media_pasajeros + df_std_pasajeros
print("Intervalo de los datos es: "+"[",delta_menos,"-",delta_mas,"]")

El promedio de pasajeros es:  280.2986111111111
La desviación estandar es:  119.96631694294321
Intervalo de los datos es: [ 160.3322941681679 - 400.2649280540543 ]


* Genere una nueva columna en la tabla de datos que se llame outlier y asignele 0.

In [40]:
df["outlier"]="0"


* Ejecuten un loop que recorra cada observación de passengers , donde si la observación se escapa de la tendencia sea 1, de lo contrario 0.

In [67]:
k=0
contador=0
for rowname,rowseries in df.iterrows():
    if (rowseries["passengers"] < delta_menos) or (rowseries["passengers"]> delta_mas):
        df.set_value(k, 'outlier', 1)
        contador+=1
    else:
        df.set_value(k, 'outlier', 0)
    k+=1
print(df)

       year      month  passengers  outlier  underperforming
0    1949.0    January       112.0      1.0              1.0
1    1949.0   February       118.0      1.0              1.0
2    1949.0      March       132.0      1.0              1.0
3    1949.0      April       129.0      1.0              1.0
4    1949.0        May       121.0      1.0              1.0
5    1949.0       June       135.0      1.0              1.0
6    1949.0       July       148.0      1.0              1.0
7    1949.0     August       148.0      1.0              1.0
8    1949.0  September       136.0      1.0              1.0
9    1949.0    October       119.0      1.0              1.0
10   1949.0   November       104.0      1.0              1.0
11   1949.0   December       118.0      1.0              1.0
12   1950.0    January       115.0      1.0              1.0
13   1950.0   February       126.0      1.0              1.0
14   1950.0      March       141.0      1.0              1.0
15   1950.0      April  

C:\Users\proveloz\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  """
C:\Users\proveloz\Anaconda3\lib\site-packages\ipykernel_launcher.py:8: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  


* ¿Cuántas observaciones se pueden clasificar como casos extremos?

In [52]:
print("La cantidad de observaciones que se escapa del intervalo"+"[",delta_menos,"-",delta_mas,"]"+" son: ", contador)

La cantidad de observaciones que se escapa del intervalo[ 160.3322941681679 - 400.2649280540543 ] son:  53


Esta cantidad se consideran extremas, ya que no se encuentran en el rango: $$280.29\pm119.96$$.
Si la cantidad son 53 datos que están fuera de este rango, podríamos comparar que 53 entre 143 se encuentran fuera de este rango, decir, cerca del 37%, no se encuentra dentro de este invervalo permitido. 